app/src/main/java/com/github/polypoly/app/network/RemoteDB.kt

```diff
@@ -105,4 +107,66 @@ open class RemoteDB(
         }.addOnFailureListener(registerPromise::completeExceptionally)
         return registerPromise
     }
+
+    override fun removeValue(key: String): CompletableFuture<Boolean> {
+        return keyExists(key).thenCompose { exists ->
+            if(exists) {
+                deleteChangeListener(key).thenCompose {
+                    setValue(key, null)
+                }
+            } else {
+                CompletableFuture.completedFuture(false)
+            }
+        }
+    }
+
+    // ========================================================================== LISTENERS
+    /**
+     * Gets the reference of a given key and if it exists, executes the given action with that reference
+     * @return a promise with true if the action was successfully executed, false else or if no value was found
+     */
+    private fun getRefAndThen(key: String, action: (DatabaseReference) -> Unit): CompletableFuture<Boolean> {
+        return keyExists(key).thenCompose { exists ->
+            if(exists) {
+                action(rootRef.child(key))
+            }
+            CompletableFuture.completedFuture(exists)
+        }
+    }
+
+    private val changeListeners = mutableMapOf<String, ValueEventListener>()
+    private val removeListeners = mutableMapOf<String, ValueEventListener>()
+
+    override fun <T : Any> addChangeListener(
```

Comment:
I don't think that such design is very safe on a large application viewpoint: if only one listener can be added per key, what if two actors want to listen a given key at the same time? Even if no such case is explicitly intended, `addChangeListener` still has the huge side effect of clearing the potentially previously stored listener without any ways of knowing it.
This looks quite error prone as it shares a responsibility across actors that do not know each others and could cause very hard to find bugs...
Maybe you could instead support a list of listeners waiting for a change, and add an optional tag when registering a new one so that the user can remove it using the same tag without affecting any other listener still pending (it's a bit more complex but much safer and flexible, and that's how firebase work as you can add any number of listeners to a single DatabaseReference = to a single key).

Reply:
Good point! I'll add a tag system so that we can have multiple listeners for a given `DataReference`